content/docs/rfcs/32/README.md

```diff
@@ -0,0 +1,560 @@
+
+1. Registration
+2. Signaling
+3. Verification and slashing
+
+
+## Registration
+
+Depending on the application requirements, the registration can be implemented in different ways, for example: 
+- centralized registrations, by using a central server
+- decentralized registrations, by using a smart contract
+
+What is important is that the user's identity commitment is stored in a merkle tree, 
+and the users can obtain a merkle proof proving that they are part of the group.
+
+Also depending on the application requirements, 
+usually a financial or social stake is introduced.
+
+An example for financial stake is: For each registration a certain amount of ETH is required.
+An example for social stake is using InterRep as a registry - 
+users need to prove that they have a highly reputable social media account.
+
+### Implementation notes
+
+#### User identity
+
+The user's identity is composed of:
+
+```
+{
+    identity_secret: [identity_nullifier, identity_trapdoor],
+    identity_secret_hash: poseidonHash(identity_secret),
+    identity_commitment: poseidonHash([identity_secret_hash])
+}
+```
+
+For registration, the user needs to submit their `identity_commitment` (along with any additional registration requirements) to the registry. 
+Upon registration, they should receive `leaf_index` value which represents their position in the merkle tree. 
+Receiving a `leaf_index` is not a hard requirement and is application specific. 
+The other way around is the users calculating the `leaf_index` themselves upon successful registration.
+
+## Signaling
+
+After registration, 
+the users can participate in the application by sending signals to the other participants in a decentralised manner or to a centralised server. 
+Along with their signal, 
+they need to generate a ZK-Proof by using the circuit with the specification described above. 
+
+For generating a proof, 
+the users need to obtain the required parameters or compute them themselves, 
+depending on the application implementation and client libraries supported by the application.
+For example the users can store the membership merkle tree on their end and 
+generate a merkle proof whenever they want to generate a signal. 
+
+### Implementation notes
+
+#### Signal hash
+
+The signal hash can be generated by hashing the raw signal (or content) using the `keccak256` hash function.
+
+#### Epoch
+
+The epoch can be generated by hashing a raw string (i.e UNIX timestamp) value using `keccak256`.
+
+####  Obtaining merkle proof
+
+The merkle proof should be obtained locally or from a trusted third party. 
+By using the [incremental merkle tree algorithm](https://github.com/appliedzkp/incrementalquintree/blob/master/ts/IncrementalQuinTree.ts), 
+the merkle can be obtained by providing the `leaf_index` of the `identity_commitment`. 
+The proof (`merkle_proof`) is composed of the following fields:
+
+```
+{
+  root: bigint
+  indices: number[]
```

Comment:
Can you please add an explanation for each of these fields?

Reply:
You may have missed this comment ☝️  https://github.com/vacp2p/rfc/pull/482/files#r814354721